### PR TITLE
(2686) Enable linking ISPF ODA with non-ODA activities via the UI

### PR DESCRIPTION
**Pull request by CristinaRO**
## Changes in this PR
- Enable activity linking for ISPF activities - manually via the user interface

### After
Screenshot 2022-11-28 at 13 39 49

Screenshot 2022-11-28 at 16 36 27

## Next steps

- [ ] Is an ADR required? An ADR should be added if this PR introduces a change to the architecture.
- [x] Is a changelog entry required? An entry should always be made in `CHANGELOG.md`, unless this PR is a small tweak which has no impact outside the development team.
- [ ] Do any environment variables need amending or adding?
- [ ] Have any changes to the XML been checked with the IATI validator? See [XML Validation](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/doc/xml-validation.md)

### CODE REVIEW

File: app/models/activity.rb [hunk 1]
```diff
@@ -621,8 +621,8 @@ def ensure_linked_activity_reciprocity
   end
 
   def linkable_activities
-    return [] unless (programme? || project?) && is_ispf_funded?
-    return [] if project? && parent.linked_activity.nil?
+    return [] unless is_ispf_funded?
```
Comment (Gweaton) [inline app/models/activity.rb:624]: [The commit](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/pull/1911/commits/de1247c0c6f0bfd5e449e127850d144bf908c2f1) message of the commit with this change is quite similar to the [previous one](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/pull/1911/commits/1018f70b911a83438c2f9e33679baf13645a7df3), and the [subsequent one](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/pull/1911/commits/0ed1856cdf90f1c297e83fa13dc1e3e8ebef37bf).

Can they be squashed together or reworded to make it a bit clearer what's going on for each?

**Review by Gweaton — APPROVED** [review]
This looks great, thanks for taking care of it all while I've been off. Just left a couple of small things that would be good to address before merging.

File: app/models/activity.rb [hunk 2]
```diff
@@ -602,6 +603,22 @@ def historic?
     programme_status.in?(["completed", "stopped", "cancelled"])
   end
 
+  def ensure_linked_activity_reciprocity
+    return unless saved_change_to_attribute?(:linked_activity_id)
+
+    formerly_linked_activity_id = saved_change_to_attribute(:linked_activity_id).first
+    if formerly_linked_activity_id && (formerly_linked_activity = Activity.find_by(id: formerly_linked_activity_id)).present?
+      # we do not want to propagate this, because it would nullify self.linked_activity_id
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/models/activity.rb:612]: > What do we mean by `propagate` here? Removing the linked activity from `self` as well?

We mean "make the formerly linked activity invoke the post-save callback". That's why we're calling `update_columns`, which doesn't trigger model callbacks. (If we allowed it to trigger callbacks, it would remove the `linked_activity_id` from the current activity, because for the formerly linked activity, that's the `formerly_linked_activity`, so it would cancel out our efforts to link it to its new "link partner".

[the remark this answers is not in the document]
Comment (CristinaRO) [reply app/models/activity.rb:612]: > Just for my own understanding, am I right in reading that this method is:
> 
> 1. Removing any formerly linked activities
> 2. Making sure the newly-linked activity's `linked_activity` is set to the current activity
> 
> ?

Yep, that's what it's doing. It made sense to us to have one method, I could be persuaded to change it but only if there was a very strong argument. >.>

File: spec/models/activity_spec.rb [hunk 3]
```diff
@@ -2204,6 +2204,55 @@ def create_report(organisation, fund)
...
+          is_oda: false,
+          linked_activity: oda_activity)
+
+        oda_activity.reload
+        expect(oda_activity.linked_activity).to eq(non_oda_activity)
+      end
+    end
+
+    context "when removing a linked activity" do
+      it "unlinks the old activity from self" do
```
[the remark this answers is not in the document]
Comment (CristinaRO) [reply spec/models/activity_spec.rb:2221]: Do you mean reword the test description, or add a new test? I thought we covered all the combinations. Could you give us a code snippet or discuss synchronously to avoid extra heapings of confusion? 😅

File: app/models/activity.rb (hunk 1 above)
Comment (CristinaRO) [reply app/models/activity.rb:624]: One of the reasons for splitting them off was ease of reviewing, so if you as a reviewer would be ok with them squished together, we can do that!

File: spec/models/activity_spec.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gweaton) [reply spec/models/activity_spec.rb:2221]: Ah, I think I understand what this means now - maybe rewording this to `it "ensures the link is removed from both sides of the relationship"` would have made more sense to me (I think I was confused by the use of `self` here.

File: app/models/activity.rb (hunk 1 above)
Comment (Gweaton) [reply app/models/activity.rb:624]: Yeah, let's do it once we're ready to merge

File: app/models/activity.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gweaton) [reply app/models/activity.rb:612]: Ok, we've agreed to leave this for now and possibly address in a future commit
